Components/CategoryResolver.php

```diff
@@ -129,12 +137,81 @@ private function removeProductsFromNotAssignedRemoteCategories(array $assignedCa
         /** @var int $currentProductCategoryId */
         foreach ($currentProductCategoryIds as $currentProductCategoryId) {
             if (!in_array($currentProductCategoryId, $assignedCategoryIds)) {
+                $this->deleteAssignmentOfLocalCategories($currentProductCategoryId, $articleId);
                 $this->productToRemoteCategoryRepository->deleteByConnectCategoryId($currentProductCategoryId, $articleId);
             }
         }
     }
 
     /**
+     * @param int $currentProductCategoryId
+     * @param int $articleId
+     */
+    private function deleteAssignmentOfLocalCategories($currentProductCategoryId, $articleId)
+    {
+        $localCategoriesIds = $this->manager->getConnection()->executeQuery(
+            'SELECT local_category_id FROM s_plugin_connect_categories_to_local_categories WHERE remote_category_id = ?',
+            [$currentProductCategoryId]
+        )->fetchAll(\PDO::FETCH_COLUMN);
+        if ($localCategoriesIds) {
+            foreach ($localCategoriesIds as $categoryId) {
+                $this->manager->getConnection()->executeQuery(
+                    'DELETE FROM `s_articles_categories` WHERE `articleID` = ? AND `categoryID` = ?',
+                    [$articleId, $categoryId]
+                );
+                $this->categoryDenormalization->removeAssignment($articleId, $categoryId);
+            }
+            $this->deleteEmptyConnectCategories($localCategoriesIds);
+        }
+    }
+
+    /**
+     * @param int[] $categoryIds
+     */
+    public function deleteEmptyConnectCategories($categoryIds)
+    {
+        foreach ($categoryIds as $categoryId) {
+            $articleCount = (int) $this->manager->getConnection()->fetchColumn(
+                'SELECT COUNT(id) FROM s_articles_categories WHERE categoryID = ?',
+                [$categoryId]
+            );
+            if ($articleCount === 0) {
+                $this->deleteEmptyCategory($categoryId);
+            }
+        }
+    }
+
+    /**
+     * @param int $categoryId
+     */
+    private function deleteEmptyCategory($categoryId)
+    {
+        $connectImported = $this->manager->getConnection()->fetchColumn(
+            'SELECT connect_imported_category FROM s_categories_attributes WHERE categoryID = ?',
+            [$categoryId]
+        );
+
+        $childCount = (int) $this->manager->getConnection()->fetchColumn(
```

Comment:
you can extract this query in private method. then in if statement below 
```if ($connectImported == 1 && $this->countChildCategories($categoryId) === 0) {```

By this way the query won't be executed when ```$categoryId``` is not imported from connect